### ISSUE TITLE
Retain CSRF token private key on Helm chart upgrades

### ISSUE THREAD

**Issue by jnoordsij** [kubernetes/dashboard#8756]
### What would you like to be added?

In the current implementation, the secret value for the CSRF secret is randomly generated on each chart upgrade, due to the value being random. To make this more consistent, it would be helpful to re-use any existing value if present.

The answer at https://stackoverflow.com/questions/67617808/helm-how-to-avoid-recreating-secrets-on-upgrade might be a good start to implement such behavior.

### Why is this needed?

Slightly less disruptive upgrade behavior.

**Comment by floreks** [comment]
This is actually a good thing that the private key will be rotated on upgrade. This is not disruptive at all for the user since it is used dynamically. The only thing that we might want to add is to rotate pods after upgrade to make sure that all of them use the same key. We can easily do that via the annotation sha.

**Comment by jnoordsij** [comment]
Yeah fair point, rotating it isn't a bad thing per se; only if inconsistent results are possible, which I saw their might. But probably this is not really much of an issue in any practical situation; feel free to close this and replace it with a more suitable suggestion!

**Comment by floreks** [comment]
I have added a custom annotation that will ensure all pods using csrf key will be rotated on chart upgrade. The chart will be released right after the linked PR gets merged.

**Comment by user81230** [comment]
Took me quite a deal to understand why does helm every time stage a change in csrf key value, but in the end, secret contains the same value as before.
Turns out, helmfile apply (or helm upgrade) outputs expected changes by templating as "helm upgrade --dry-run=client" but there are no changes when actually applying. The default behavior of "lookup" function used in templates/_helpers.tpl is so that it always returns empty map when dry-running (e.g. templating, diff), and thus it always returns different results between dry running and applying.
Also I found that they've changed it in Helm v3.13 by adding --dry-run={client or server} choice, and it seems that helm diff plugin hasn't supported this yet.
Thanks.
